subsys/net/ip/tcp2.c

```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
 	return len;
 }
 
-static void tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
+static int tcp_finalize_pkt(struct net_pkt *pkt)
 {
-	struct net_buf *buf = net_pkt_get_frag(pkt, K_NO_WAIT);
-	struct tcphdr th;
-
-	memset(&th, 0, sizeof(th));
-
-	th.th_sport = conn->src->sin.sin_port;
-	th.th_dport = conn->dst->sin.sin_port;
-
-	th.th_off = 5;
-	th.th_flags = flags;
-	th.th_win = htons(conn->win);
-	th.th_seq = htonl(conn->seq);
+	net_pkt_cursor_init(pkt);
 
-	if (ACK & flags) {
-		th.th_ack = htonl(conn->ack);
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_ipv4_finalize(pkt, IPPROTO_TCP);
 	}
 
-	memcpy(net_buf_add(buf, sizeof(th)), &th, sizeof(th));
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_ipv6_finalize(pkt, IPPROTO_TCP);
+	}
 
-	net_pkt_frag_insert(pkt, buf);
+	return -EINVAL;
 }
 
-static void ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
```

Comment:
call it tcp_create() to stay with the same naming we have all over the net stack.